src/main/java/io/temporal/failure/ApplicationFailure.java

```diff
@@ -51,49 +51,47 @@
  */
 public final class ApplicationFailure extends TemporalFailure {
   private final String type;
-  private final Value details;
+  private final Values details;
   private boolean nonRetryable;
 
   /**
+   * New ApplicationFailure with {@link #isNonRetryable()} flag set to false. Note that this
+   * exception still can be retried by the service if its type is included into doNotRetry property
+   * of the correspondent retry policy.
+   *
    * @param message optional error message
    * @param type optional error type that is used by {@link
    *     io.temporal.common.RetryOptions#addDoNotRetry(String...)}.
    * @param details optional details about the failure. They are serialized using the same approach
-   *     as arguments and results and can be accessed through {@link #getDetails()}
-   * @param cause failure cause. Each element of the cause chain is converted to ApplicationFailure
-   *     if it doesn't extend {@link TemporalFailure}.
+   *     as arguments and results.
    */
-  public ApplicationFailure(String message, String type, Object details, Exception cause) {
-    this(message, type, new EncodedValue(details), false, cause);
+  public static ApplicationFailure newFailure(String message, String type, Object... details) {
```

Comment:
why not just call it newRetryableFailure?

Reply:
I called it initially. But then I found that it is very confusing that it can be not retried if the policy says so.